### PR TITLE
Add macro documentation.

### PR DESCRIPTION
**Pull request by de-vri-es**
This PR adds documentation for the `interface!` macro.

The `macros` feature is not enabled by default, so to read the docs in a browser, run: `cargo doc --all-features --open`

### CODE REVIEW

File: src/macros.rs [hunk 1]
```diff
@@ -1,10 +1,305 @@
...
+///     interface supermarket {
+///         /// Greet the cashier.
+///         ///
+///         /// The cashier will reply with their own greeting.
+///         service 1 greet_cashier: String -> String,
+///
+///         /// Purchase tomatoes.
+///         ///
+///         /// The response of the cashier depends on the update messages exchanged.
+///         /// If you run away after they have sent the `price`, or if you pay with a nun-fungable token,
```
Comment (vcarpani) [inline src/macros.rs:133]: I like tokens that don't become mushrooms, but I guess it is non-fungible

File: src/macros.rs [hunk 2]
```diff
@@ -1,10 +1,305 @@
...
+///             request_update 2 run_away: (),
+///         }
+///
+///         /// Mutter something as you're walking through the supermarket.
+///         ///
+///         /// Since no-one will respond, this is a stream rather than a service.
+///         ///
+///         /// Popular phrases include:
+///         ///  * Woah thats cheap!
+///         ///  * Everything used to be better in the old days...
```
Comment (vcarpani) [inline src/macros.rs:154]: the good old days

File: src/macros.rs (hunk 1 above)
Comment (de-vri-es) [reply src/macros.rs:133]: Hmm, probably. Although either work equally as well as a money substitute.

File: src/macros.rs (hunk 2 above)
Comment (de-vri-es) [reply src/macros.rs]: Yes, good point. Updated.

File: macros/src/interface/generate.rs [hunk 3]
```diff
@@ -72,17 +74,6 @@ fn generate_client(item_tokens: &mut TokenStream, fizyr_rpc: &syn::Ident, interf
...
-			///
-			/// See [`fizyr_rpc::Peer::connect`](https://docs.rs/fizyr-rpc/latest/fizyr_rpc/struct.Peer.html#method.connect) for more details.
-			pub async fn connect<'a, Transport, Address>(address: Address, config: Transport::Config) -> std::io::Result<Self>
-			where
-				Address: 'a,
-				Transport: #fizyr_rpc::transport::Transport<Body = F::Body> + #fizyr_rpc::util::Connect<'a, Address>,
-			{
-				Ok(#fizyr_rpc::Peer::<Transport>::connect(address, config).await?.into())
-			}
-
```
Comment (hgaiser) [inline macros/src/interface/generate.rs:85]: Why is this removed?

File: src/macros.rs [hunk 4]
```diff
@@ -1,10 +1,305 @@
+//! Extra macro documentation.
+
 #[doc(hidden)]
 pub use fizyr_rpc_macros::interface as interface_impl;
 
 #[macro_export]
 /// Define an RPC interface.
+///
+/// This macro generates a module with a `Client` and `Server` struct,
+/// and some more helper types.
```
Comment (hgaiser) [inline src/macros.rs:10]: Maybe briefly add what these helper types are intended for? Or is it so small that it isn't noteworthy?

File: src/macros.rs [hunk 5]
```diff
@@ -1,10 +1,305 @@
...
+///
+/// The client struct can be used to initiate requests and send stream messages.
```
Comment (hgaiser) [inline src/macros.rs:12]:
```suggestion
/// The client struct can be used to initiate requests and stream messages.
```
Or is this conceptually very different?

File: src/macros.rs [hunk 6]
```diff
@@ -1,10 +1,305 @@
...
+/// It can be created from a [`PeerWriteHandle`] or a [`PeerHandle`].
+/// Note that if you create the client from a [`PeerHandle`], the [`PeerReadHandle`] will not be accessible.
```
Comment (hgaiser) [inline src/macros.rs:14]: Might be interesting to say why, unless it is too complicated of course.

File: src/macros.rs [hunk 7]
```diff
@@ -1,10 +1,305 @@
...
+///
+/// For the server struct it is exacatly opposite: it can be used to receive requests and stream messages.
```
Comment (hgaiser) [inline src/macros.rs:16]:
```suggestion
/// For the server struct it is exactly the opposite: it can be used to receive requests and stream messages.
```

File: src/macros.rs [hunk 8]
```diff
@@ -1,10 +1,305 @@
...
+/// It can be created from a [`PeerReadHandle`] or a [`PeerHandle`],
+/// but creating it from a [`PeerHandle`] will discard the [`PeerWriteHandle`].
```
Comment (hgaiser) [inline src/macros.rs:18]: Because servers can't initiate interactions, right?

File: src/macros.rs [hunk 9]
```diff
@@ -1,10 +1,305 @@
...
+/// fizyr_rpc::interface! {
+///     // The `interface` keyword defines an RPC interface.
+///     // You must have exactly one interface definition in the macro invocation.
+///     //
+///     // The macro generates a module with the same name containing the generated types.
+///     // You can adjust the visiblity of the generated module with the `pub` keyword as normal.
+///     // You can use this to take full control over the public module structure,
+///     // by declaring the interface private and re-exporting the module contents from a different location.
+///     //
+///     // Each item in can have user written documentation.
```
Comment (hgaiser) [inline src/macros.rs:37]:
```suggestion
///     // Each item can have user written documentation.
```

File: src/macros.rs [hunk 10]
```diff
@@ -1,10 +1,305 @@
...
+///     // Simply write doc comments with triple slashes as usual.
+///     // This applies to the interface definition, a service definitions, update definitions and stream definitions.
+///     pub interface $interface_name {
```
Comment (hgaiser) [inline src/macros.rs:40]: Maybe write somewhere that `$interface_name` needs to be replaced.

File: src/macros.rs [hunk 11]
```diff
@@ -1,10 +1,305 @@
...
+///         // The `service` keyword defines a service.
+///         //
+///         // You can have any amount of service definition inside an interface definition.
```
Comment (hgaiser) [inline src/macros.rs:43]:
```suggestion
///         // You can have any amount of service definitions inside an interface definition.
```

File: src/macros.rs [hunk 12]
```diff
@@ -1,10 +1,305 @@
...
+///         //
+///         // The $id is used as the service ID and must be an i32.
+///         // The ID must be unique for all services in the interface.
```
Comment (hgaiser) [inline src/macros.rs:46]: What happens if I make a service with non-unique ID?

File: src/macros.rs [hunk 13]
```diff
@@ -1,10 +1,305 @@
...
+///         //
+///         // The $name is the name of the service.
+///         // It is used to generate function en type names.
```
Comment (hgaiser) [inline src/macros.rs:49]:
```suggestion
///         // It is used to generate function and type names.
```

You Dutchman :p

File: src/macros.rs [hunk 14]
```diff
@@ -1,10 +1,305 @@
...
+///         // It must be a valid Rust identifier and should be lower-case with underscores.
```
Comment (hgaiser) [inline src/macros.rs:50]:
```suggestion
///         // It must be a valid Rust identifier and should be lowercase with underscores.
```

File: src/macros.rs [hunk 15]
```diff
@@ -1,10 +1,305 @@
...
+///         //
+///         // The $request_type and $response_type indicate the message body for the request and the response.
+///         // If there is no data in a request or response, you can use the unit type: `()`
+///         service $id $name: $request_type -> $response_type,
+///
+///         // If a service has update messages, you can declare them in service a block.
```
Comment (hgaiser) [inline src/macros.rs:56]:
```suggestion
///         // If a service has update messages, you can declare them in a block.
```

File: src/macros.rs [hunk 16]
```diff
@@ -1,10 +1,305 @@
...
+///         service $id $name: $request_type -> $response_type {
+///             // The `request_update` keyword defines a request update.
+///             // You can have any amount of request updates inside a service definition.
+///             //
+///             // The $id is used as the service ID for the update and must be an i32.
+///             // The ID must be unique for all request updates in the service.
+///             //
+///             // The $name is the name of the update message.
+///             // It is used to generate function en type names.
```
Comment (hgaiser) [inline src/macros.rs:65]:
```suggestion
///             // It is used to generate function and type names.
```

File: src/macros.rs [hunk 17]
```diff
@@ -1,10 +1,305 @@
...
+///             // It must be a valid Rust identifier and should be lower-case with underscores.
```
Comment (hgaiser) [inline src/macros.rs:66]:
```suggestion
///             // It must be a valid Rust identifier and should be lowercase with underscores.
```

File: src/macros.rs [hunk 18]
```diff
@@ -1,10 +1,305 @@
...
+///             request_update $id $name: $body_type,
+///
+///             // The `response_update` keyword defines a response update.
+///             // You can have any amount of response updates inside a service definition.
+///             //
+///             // The $id is used as the service ID for the update and must be an i32.
+///             // The ID must be unique for all response updates in the service.
+///             //
+///             // The $name is the name of the update message.
+///             // It is used to generate function en type names.
```
Comment (hgaiser) [inline src/macros.rs:79]:
```suggestion
///             // It is used to generate function and type names.
```

File: src/macros.rs [hunk 19]
```diff
@@ -1,10 +1,305 @@
...
+///             // It must be a valid Rust identifier and should be lower-case with underscores.
```
Comment (hgaiser) [inline src/macros.rs:80]:
```suggestion
///             // It must be a valid Rust identifier and should be lowercase with underscores.
```

File: src/macros.rs [hunk 20]
```diff
@@ -1,10 +1,305 @@
...
+///         }
+///
+///         // The `stream` keyword defines a stream message.
+///         // You can have any amount of stream definitions in an interface definition.
+///         //
+///         // The $id is used as the service ID of the stream message and must be an i32.
+///         // The ID must be unique for all streams in the interface.
+///         //
+///         // The $name is the name of the stream.
+///         // It is used to generate function en type names.
```
Comment (hgaiser) [inline src/macros.rs:94]:
```suggestion
///         // It is used to generate function and type names.
```

File: src/macros.rs [hunk 21]
```diff
@@ -1,10 +1,305 @@
...
+///         // It must be a valid Rust identifier and should be lower-case with underscores.
```
Comment (hgaiser) [inline src/macros.rs:95]:
```suggestion
///         // It must be a valid Rust identifier and should be lowercase with underscores.
```

File: src/macros.rs [hunk 22]
```diff
@@ -1,10 +1,305 @@
...
+/// [`PeerReadHandle`]: crate::PeerReadHandle
 macro_rules! interface {
 	($($tokens:tt)*) => {
 		$crate::macros::interface_impl!{$crate; $($tokens)*}
 	}
 }
+
+/// Example module for the `interface!` macro.
+///
+/// You can compare the source the the gerated documentation to inspect the generated API.
```
Comment (hgaiser) [inline src/macros.rs:115]:
```suggestion
/// You can compare the source of the generated documentation to inspect the generated API.
```

File: src/macros.rs [hunk 23]
```diff
@@ -1,10 +1,305 @@
...
+///             request_update 2 run_away: (),
+///         }
+///
+///         /// Mutter something as you're walking through the supermarket.
+///         ///
+///         /// Since no-one will respond, this is a stream rather than a service.
+///         ///
+///         /// Popular phrases include:
+///         ///  * Woah thats cheap!
+///         ///  * Everything used to be better in the good old days...
```
Comment (hgaiser) [inline src/macros.rs:154]: xD

true though.

**Review by hgaiser — COMMENTED** [review]
General remark, try to avoid "can be" in sentences if it doesn't add anything. For example:

```
 /// The client struct can be used to initiate requests and send stream messages.
```

Versus

```
 /// The client struct is used to initiate requests and send stream messages.
```

Bit nitpicky, but looks better IMO.

File: src/macros.rs (hunk 4 above)
Comment (de-vri-es) [reply src/macros.rs:10]: No, it's more that there are so many, and they're subdivided into multiple modules =]

File: macros/src/interface/generate.rs (hunk 3 above)
Comment (de-vri-es) [reply macros/src/interface/generate.rs:85]: Because it's complicated to use correctly. You would always need to manually fill in the `Transport` and `Address` type when calling it, which also requires you to fill in the lifetime.

I actually discussed this with you, and we decided to drop the `connect()` function :p

File: src/macros.rs (hunk 5 above)
Comment (de-vri-es) [reply src/macros.rs:12]: Yes, it's different, you don't initiate stream messages, you just send them and then you're done :)

File: src/macros.rs (hunk 6 above)
Comment (de-vri-es) [reply src/macros.rs:14]: Well, it's just a consequence of how the API works. There's no specific reason other than there is no logical way to expose the read handle when consuming the `PeerHandle`.

File: src/macros.rs (hunk 8 above)
Comment (de-vri-es) [reply src/macros.rs:18]: Well, that depends on your terminology. The listener can accept connections and then initiate requests. But  the generated Server can only handle requests and stream messages.

File: src/macros.rs (hunk 10 above)
Comment (de-vri-es) [reply src/macros.rs:40]: I should expain what `$interface_name` is used for like with the other keywords, yeah.

File: src/macros.rs (hunk 12 above)
Comment (de-vri-es) [reply src/macros.rs:46]: You'll get a very beautiful error :) Try it :D

File: src/macros.rs (hunk 15 above)
Comment (de-vri-es) [reply src/macros.rs:56]: Hmm, how about "the service block"?

**Comment by de-vri-es** [comment]
> General remark, try to avoid "can be" in sentences if it doesn't add anything

I changed it for "The ... struct {can be => is} used to ...", but I kept it for `It can be created from ...` because there are multiple options.

File: src/macros.rs (hunk 15 above)
Comment (hgaiser) [reply src/macros.rs:57]: Works for me :)

**Comment by hgaiser** [comment]
> I changed it for "The ... struct {can be => is} used to ...", but I kept it for `It can be created from ...` because there are multiple options.

Sounds good to me.

File: src/macros.rs (hunk 10 above)
Comment (de-vri-es) [reply src/macros.rs:41]: Reworded it now as a bit more formal parameter description, like the rest.
